File: Problem3/AccountOverdrawSafeDemo.java
class Account {
  String name; //name of joint account
  int balance; // the balance that is present

  Account(String name, int balance) //constructor to initialize
  {
    this.balance = balance;
    this.name = name;
  }

  synchronized void withdraw(String name, int with_draw) // method to withdraw cash but not getting into negative balance
  {
    if (balance >= with_draw) {
      balance = balance - with_draw;
      System.out.println(name + " has withdraw : " + with_draw + "\nYour balance : " + balance + "\n");
      try {
        Thread.sleep(50);
      } catch (Exception exception) {}
    } else {
      System.out.println(name + ", You don't have " + with_draw + "\nYour balance : " + balance + "\n");
      try {
        Thread.sleep(50);
      } catch (Exception exception) {}
    }
  }

}

class AccountOverdrawSafeDemo extends Thread //a class to call withdraw method 
{
  Account account;
  String name; //name of varible which is withdrawing the amount

  AccountOverdrawSafeDemo(Account account, String name) //constructor to initialize
  {
    this.account = account;
    this.name = name;
  }

  public void run() //overridding run method of thread class
  {
    int index = 0;
    while (index <= 5) {
      account.withdraw(name, (int)(Math.random() * 1000) + 1); // randomly deducting amount
      index++;
    }
  }

}
class Main {
  public static void main(String[] args) {
    Account account = new Account("Ram & Sita", 1000);
    AccountOverdrawSafeDemo thread1 = new AccountOverdrawSafeDemo(account, "Ram");
    AccountOverdrawSafeDemo thread2 = new AccountOverdrawSafeDemo(account, "Sita");
    thread1.start();
    thread2.start();
  }
}

File: Problem3/AccountOverdrawSafeDemo.js
//creating account class
class Account {
  constructor(name1, name2, balance) {
    this.name1 = name1;
    this.name2 = name2;
    this.balance = balance;
  }
  // method to withdraw amount from account
  async withdraw(amount, name) {
    //only if amount is less then amount will withdraw
    if (this.balance >= amount) {
      console.log(`${name} is withdrawing ${amount}`);
      await new Promise(resolve => {
        this.balance -= amount;
        console.log(`New balance: ${this.balance}`); // withdraw succesful
        setTimeout(resolve, 10000);
      }
      );
      return true;
    } else {
      //if they don't have balance
      console.log(`${name} you do not have ${amount}`);
      return false;
    }
  }
}

class AccountOverdrawSafeDemo {
  //run method to run tasks or threads 
  static async run() {
    //joint account holder's, account creation 
    const account = new Account('Ishaan', 'Akar', 1000);
    let flag = true; // flag to alternate between thread1 and thread2

    // Withdraw 1
    const thread1 = async (name) => {
      while (true) {
        //  delay to decrease the collapse 
        await new Promise(resolve => setTimeout(resolve, 1000));
        //randomly generating withdraw amount

        const flag = await account.withdraw(Math.floor(Math.random() * 500) + 100, name); //calling method
        if (!flag) break;
      }

    };
    // Withdraw 2
    const thread2 = async (name) => {
      while (true) {
        //  delay to decrease the collapse 
        await new Promise(resolve => setTimeout(resolve, 4000));
        //randomly generating withdraw amount

        const flag = await account.withdraw(Math.floor(Math.random() * 500) + 100, name); //calling method
        if (!flag) break;
      }
    };

    await Promise.all([thread1('Ishaan'), thread2('Akar')]);
    console.log(`Final balance: ${account.balance}`);
  }
}


// Test AccountOverdrawSafeDemo
console.log('Running AccountOverdrawSafeDemo...\n');
AccountOverdrawSafeDemo.run().catch(console.error);
